source/common/buffer/watermark_buffer.h

```diff
@@ -39,6 +39,9 @@ class WatermarkBuffer : public Instance {
     return wrapped_buffer_->search(data, size, start);
   }
   int write(int fd) override;
+  Event::Libevent::BufferPtr& buffer() override {
+    return static_cast<LibEventInstance&>(*wrapped_buffer_).buffer();
```

Comment:
@htuch as discussed this solves the problem where move didn't work in both directions (now tested) because OwnedImpl assumed it was moving another OwnedImpl.  The down-side is that WatermarkBuffer now only can do move()s if it wraps an OwnedImpl.  That said the Envoy code base already assumes all buffers are OwnedImpls so I don't think we're losing any flexibility, it's just ugly from a purist perspective.